builtin/game/falling.lua

```diff
@@ -243,14 +257,26 @@ function core.check_single_for_falling(p)
 				core.get_node_max_level(p_bottom))) and
 
 				(not d_bottom.walkable or d_bottom.buildable_to) then
-			convert_to_falling_node(p, n)
-			return true
+			return convert_to_falling_node(p, n)
 		end
 	end
 
 	if core.get_item_group(n.name, "attached_node") ~= 0 then
+		local def = core.registered_nodes[n.name]
 		if not builtin_shared.check_attached_node(p, n) then
+			if def and def.on_detach then
+				local ok = def.on_detach(table.copy(p), table.copy(n))
+				if type(ok) == "boolean" then
+					return ok
+				end
+			end
+			local oldmeta = core.get_meta(p):to_table().fields
+
 			drop_attached_node(p)
+
+			if def and def.after_detach then
```

Comment:
L267 and here: `def` cannot be `nil`. The `core.get_item_group ... ~= 0` always requires a valid node def.